src/core/unit.c

```diff
@@ -689,38 +689,45 @@ static void unit_remove_transient(Unit *u) {
         }
 }
 
-static void unit_free_requires_mounts_for(Unit *u) {
+static void unit_free_mounts_for(Unit *u) {
         assert(u);
 
-        for (;;) {
-                _cleanup_free_ char *path = NULL;
-
-                path = hashmap_steal_first_key(u->requires_mounts_for);
-                if (!path)
-                        break;
-                else {
-                        char s[strlen(path) + 1];
-
-                        PATH_FOREACH_PREFIX_MORE(s, path) {
-                                char *y;
-                                Set *x;
+        for (UnitMountDependencyType t = 0; t < _UNIT_MOUNT_DEPENDENCY_TYPE_MAX; ++t) {
+                for (;;) {
+                        _cleanup_free_ char *path = NULL;
 
-                                x = hashmap_get2(u->manager->units_requiring_mounts_for, s, (void**) &y);
-                                if (!x)
-                                        continue;
+                        path = hashmap_steal_first_key(u->mounts_for[t]);
+                        if (!path)
+                                break;
+                        else {
```

Comment:
if you touch this, maybe drop this indentation level (i.e. `else {}`), the first `if` branch has `break` anyway, hence the indentation is unnecessary